File: python/reddit_scraper/scraper.py
# ====================================================================
#                 REDDIT SCRAPER CORE & CLI ENGINE
# ====================================================================
#  Author      : ZetaGo-Aurum
#  GitHub      : https://github.com/ZetaGo-Aurum
#  Repository  : https://github.com/ZetaGo-Aurum/reddit-scraper
#  License     : MIT
#
#  [NOTICE & WATERMARK]
#  DO NOT REMOVE THIS WATERMARK OR AUTHOR CREDITS!
#  This software is created and maintained by ZetaGo-Aurum.
#  All rights reserved. Unauthorized removal of this header is prohibited.
# ====================================================================

import csv
import json
from typing import List, Optional, Dict, Any, Union
from .models import Post, Comment, SubredditInfo, UserProfile
from .client import RedditClient


class RedditScraper:
    def __init__(self, client: Optional[RedditClient] = None):
        self.client = client or RedditClient()

    # ---------------------------------------------------------
    # 1. Search Scraping
    # ---------------------------------------------------------
    def search(
        self,
        query: str,
        subreddit: Optional[str] = None,
        sort: str = "relevance",  # relevance, hot, top, new, comments
        time_filter: str = "all",  # all, hour, day, week, month, year
        limit: int = 25,
    ) -> List[Post]:
        """
        Search Reddit for posts matching query.
        Optionally restrict to a specific subreddit.
        """
        params = {
            "q": query,
            "sort": sort,
            "t": time_filter,
            "limit": min(limit, 100),
        }
        if subreddit:
            path = f"/r/{subreddit}/search"
            params["restrict_sr"] = "on"
        else:
            path = "/search"

        data = self.client.get(path, params=params)
        posts: List[Post] = []
        children = data.get("data", {}).get("children", [])
        for item in children:
            if item.get("kind") == "t3":
                posts.append(Post.from_dict(item))
        return posts[:limit]

    # ---------------------------------------------------------
    # 2. Subreddit Scraping
    # ---------------------------------------------------------
    def get_subreddit_posts(
        self,
        subreddit: str,
        sort: str = "hot",  # hot, new, top, rising
        time_filter: str = "all",  # all, day, week, month, year (for top)
        limit: int = 25,
    ) -> List[Post]:
        """Scrape posts from a subreddit"""
        path = f"/r/{subreddit}/{sort}"
        params = {
            "limit": min(limit, 100),
        }
        if sort == "top":
            params["t"] = time_filter

        data = self.client.get(path, params=params)
        posts: List[Post] = []
        children = data.get("data", {}).get("children", [])
        for item in children:
            if item.get("kind") == "t3":
                posts.append(Post.from_dict(item))
        return posts[:limit]

    def get_subreddit_about(self, subreddit: str) -> SubredditInfo:
        """Get subreddit metadata, rules, subscriber count, description"""
        path = f"/r/{subreddit}/about"
        data = self.client.get(path)
        return SubredditInfo.from_dict(data)

    def get_post_with_comments(
        self,
        post_id_or_url: str,
        sort: str = "confidence",  # confidence, top, new, controversial, old
        limit: int = 50,
        depth: int = 3,
    ) -> Post:
        """
        Get full post details along with its comment tree.
        Accepts post ID (e.g. '1c1abcd') or full permalink.
        """
        if post_id_or_url.startswith("http") or "/r/" in post_id_or_url:
            # extract path
            parts = post_id_or_url.split("reddit.com")[-1].split("?")[0]
            path = parts.rstrip("/")
        else:
            path = f"/comments/{post_id_or_url}"

        params = {
            "sort": sort,
            "limit": limit,
            "depth": depth,
        }
        data = self.client.get(path, params=params)

        # Reddit returns a list of 2 items: [post_listing, comment_listing]
        if not isinstance(data, list) or len(data) < 2:
            raise ValueError(f"Unexpected response format for comments: {data}")

        post_data = data[0]["data"]["children"][0]
        post = Post.from_dict(post_data)

        comment_children = data[1]["data"]["children"]
        comments: List[Comment] = []
        for c in comment_children:
            parsed = Comment.from_dict(c, depth=0)
            if parsed:
                comments.append(parsed)

        post.comments = comments
        return post

    # ---------------------------------------------------------
    # 3. User Profile Scraping
    # ---------------------------------------------------------
    def get_user_profile(self, username: str) -> UserProfile:
        """Get user karma, created date, bio, avatar"""
        # Clean username
        user = username.replace("u/", "").replace("/u/", "").strip()
        path = f"/user/{user}/about"
        data = self.client.get(path)
        return UserProfile.from_dict(data)

    def get_user_posts(
        self,
        username: str,
        sort: str = "new",  # new, hot, top
        limit: int = 25,
    ) -> List[Post]:
        """Scrape submitted posts by a user"""
        user = username.replace("u/", "").replace("/u/", "").strip()
        path = f"/user/{user}/submitted"
        params = {"sort": sort, "limit": min(limit, 100)}
        data = self.client.get(path, params=params)
        posts: List[Post] = []
        children = data.get("data", {}).get("children", [])
        for item in children:
            if item.get("kind") == "t3":
                posts.append(Post.from_dict(item))
        return posts[:limit]

    def get_user_comments(
        self,
        username: str,
        sort: str = "new",
        limit: int = 25,
    ) -> List[Comment]:
        """Scrape comments by a user"""
        user = username.replace("u/", "").replace("/u/", "").strip()
        path = f"/user/{user}/comments"
        params = {"sort": sort, "limit": min(limit, 100)}
        data = self.client.get(path, params=params)
        comments: List[Comment] = []
        children = data.get("data", {}).get("children", [])
        for item in children:
            if item.get("kind") == "t1":
                c = Comment.from_dict(item)
                if c:
                    comments.append(c)
        return comments[:limit]

    # ---------------------------------------------------------
    # Export Helpers
    # ---------------------------------------------------------
    @staticmethod
    def export_json(data: Union[List[Any], Any], filepath: str):
        """Export scraped objects to JSON file"""
        if isinstance(data, list):
            dumpable = [item.model_dump() if hasattr(item, "model_dump") else item for item in data]
        else:
            dumpable = data.model_dump() if hasattr(data, "model_dump") else data

        with open(filepath, "w", encoding="utf-8") as f:
            json.dump(dumpable, f, indent=2, ensure_ascii=False)

    @staticmethod
    def export_posts_csv(posts: List[Post], filepath: str):
        """Export posts to CSV file"""
        fields = [
            "id", "title", "author", "subreddit", "score", "num_comments",
            "created_utc", "url", "permalink", "is_self", "link_flair_text"
        ]
        with open(filepath, "w", encoding="utf-8", newline="") as f:
            writer = csv.DictWriter(f, fieldnames=fields)
            writer.writeheader()
            for p in posts:
                row = {k: getattr(p, k) for k in fields}
                writer.writerow(row)

File: lib/scraper.js
/**
 * ====================================================================
 *                 REDDIT SCRAPER CORE & CLI ENGINE
 * ====================================================================
 *  Author      : ZetaGo-Aurum
 *  GitHub      : https://github.com/ZetaGo-Aurum
 *  Repository  : https://github.com/ZetaGo-Aurum/reddit-scraper
 *  License     : MIT
 *
 *  [NOTICE & WATERMARK]
 *  DO NOT REMOVE THIS WATERMARK OR AUTHOR CREDITS!
 *  This software is created and maintained by ZetaGo-Aurum.
 *  All rights reserved. Unauthorized removal of this header is prohibited.
 * ====================================================================
 */

'use strict';

const { RedditClient } = require('./client');
const { Post, Comment, SubredditInfo, UserProfile } = require('./models');
const { exportToJson, exportToCsv, exportToMarkdown } = require('./utils');

class RedditScraper {
  constructor(clientOrConfig = null) {
    if (clientOrConfig instanceof RedditClient) {
      this.client = clientOrConfig;
    } else {
      this.client = new RedditClient(clientOrConfig);
    }
  }

  /**
   * Search Reddit posts across all Reddit or restricted to a specific subreddit.
   * 
   * @param {Object} options
   * @param {string} options.query - Search query string
   * @param {string} [options.subreddit] - Optional subreddit to restrict search to
   * @param {'relevance'|'hot'|'top'|'new'|'comments'} [options.sort='relevance']
   * @param {'all'|'hour'|'day'|'week'|'month'|'year'} [options.timeFilter='all']
   * @param {number} [options.limit=25]
   * @param {string} [options.after]
   * @returns {Promise<Post[]>}
   */
  async search({
    query,
    subreddit = null,
    sort = 'relevance',
    timeFilter = 'all',
    limit = 25,
    after = null,
  }) {
    if (!query) throw new Error('Search query is required');

    const params = {
      q: query,
      sort,
      t: timeFilter,
      limit: Math.min(limit, 100),
    };
    if (after) params.after = after;

    let path = '/search';
    if (subreddit) {
      const cleanSub = subreddit.replace(/^r\//, '');
      path = `/r/${cleanSub}/search`;
      params.restrict_sr = 'on';
    }

    const data = await this.client.get(path, params);
    const posts = [];
    const children = data?.data?.children || [];

    for (const item of children) {
      if (item.kind === 't3') {
        posts.push(Post.fromRedditJson(item));
      }
    }

    return posts.slice(0, limit);
  }

  /**
   * Scrape posts from a subreddit (hot, new, top, rising).
   * 
   * @param {Object} options
   * @param {string} options.subreddit - Name of subreddit
   * @param {'hot'|'new'|'top'|'rising'} [options.sort='hot']
   * @param {'all'|'day'|'week'|'month'|'year'} [options.timeFilter='all']
   * @param {number} [options.limit=25]
   * @param {string} [options.after]
   * @returns {Promise<Post[]>}
   */
  async getSubredditPosts({
    subreddit,
    sort = 'hot',
    timeFilter = 'all',
    limit = 25,
    after = null,
  }) {
    if (!subreddit) throw new Error('Subreddit name is required');
    const cleanSub = subreddit.replace(/^r\//, '');

    const path = `/r/${cleanSub}/${sort}`;
    const params = {
      limit: Math.min(limit, 100),
    };
    if (sort === 'top') {
      params.t = timeFilter;
    }
    if (after) params.after = after;

    const data = await this.client.get(path, params);
    const posts = [];
    const children = data?.data?.children || [];

    for (const item of children) {
      if (item.kind === 't3') {
        posts.push(Post.fromRedditJson(item));
      }
    }

    return posts.slice(0, limit);
  }

  /**
   * Get metadata, description, rules, and subscriber metrics for a subreddit.
   * 
   * @param {string} subreddit - Name of subreddit
   * @returns {Promise<SubredditInfo>}
   */
  async getSubredditAbout(subreddit) {
    if (!subreddit) throw new Error('Subreddit name is required');
    const cleanSub = subreddit.replace(/^r\//, '');
    const path = `/r/${cleanSub}/about`;

    const data = await this.client.get(path);
    return SubredditInfo.fromRedditJson(data);
  }

  /**
   * Scrape a full Reddit post and its nested comment tree.
   * 
   * @param {string} postIdOrUrl - Reddit Post ID (e.g. '1c1abcd') or full permalink / URL
   * @param {Object} [options]
   * @param {'confidence'|'top'|'new'|'controversial'|'old'} [options.sort='confidence']
   * @param {number} [options.limit=50]
   * @param {number} [options.depth=3]
   * @returns {Promise<Post>}
   */
  async getPost(postIdOrUrl, { sort = 'confidence', limit = 50, depth = 3 } = {}) {
    if (!postIdOrUrl) throw new Error('Post ID or URL is required');

    let path = '';
    if (postIdOrUrl.includes('reddit.com') || postIdOrUrl.includes('/r/')) {
      const clean = postIdOrUrl.split('reddit.com').pop().split('?')[0];
      path = clean.replace(/\/$/, '');
    } else {
      path = `/comments/${postIdOrUrl}`;
    }

    const params = {
      sort,
      limit,
      depth,
    };

    const data = await this.client.get(path, params);
    if (!Array.isArray(data) || data.length === 0) {
      throw new Error('Invalid post response returned by Reddit');
    }

    // data[0] contains post details
    const postListing = data[0]?.data?.children || [];
    if (postListing.length === 0) {
      throw new Error('Post not found');
    }
    const post = Post.fromRedditJson(postListing[0]);

    // data[1] contains comments tree
    if (data.length > 1 && data[1]?.data?.children) {
      const commentItems = data[1].data.children;
      for (const item of commentItems) {
        const comment = Comment.fromRedditJson(item, 0);
        if (comment) {
          post.comments.push(comment);
        }
      }
    }

    return post;
  }

  /**
   * Scrape public user profile, karma metrics, and account status.
   * 
   * @param {string} username - Reddit username (without u/)
   * @returns {Promise<UserProfile>}
   */
  async getUserProfile(username) {
    if (!username) throw new Error('Username is required');
    const cleanUser = username.replace(/^u\//, '');
    const path = `/user/${cleanUser}/about`;

    const data = await this.client.get(path);
    return UserProfile.fromRedditJson(data);
  }

  /**
   * Scrape submitted posts by a user.
   * 
   * @param {string} username - Reddit username
   * @param {Object} [options]
   * @param {'new'|'hot'|'top'} [options.sort='new']
   * @param {number} [options.limit=25]
   * @returns {Promise<Post[]>}
   */
  async getUserPosts(username, { sort = 'new', limit = 25 } = {}) {
    if (!username) throw new Error('Username is required');
    const cleanUser = username.replace(/^u\//, '');
    const path = `/user/${cleanUser}/submitted`;

    const data = await this.client.get(path, { sort, limit: Math.min(limit, 100) });
    const posts = [];
    const children = data?.data?.children || [];

    for (const item of children) {
      if (item.kind === 't3') {
        posts.push(Post.fromRedditJson(item));
      }
    }

    return posts.slice(0, limit);
  }

  /**
   * Scrape user comments.
   * 
   * @param {string} username - Reddit username
   * @param {Object} [options]
   * @param {'new'|'hot'|'top'} [options.sort='new']
   * @param {number} [options.limit=25]
   * @returns {Promise<Comment[]>}
   */
  async getUserComments(username, { sort = 'new', limit = 25 } = {}) {
    if (!username) throw new Error('Username is required');
    const cleanUser = username.replace(/^u\//, '');
    const path = `/user/${cleanUser}/comments`;

    const data = await this.client.get(path, { sort, limit: Math.min(limit, 100) });
    const comments = [];
    const children = data?.data?.children || [];

    for (const item of children) {
      if (item.kind === 't1') {
        const c = Comment.fromRedditJson(item, 0);
        if (c) comments.push(c);
      }
    }

    return comments.slice(0, limit);
  }

  /**
   * Utility helper to export scraped data
   */
  export(data, { format = 'json', filePath, title } = {}) {
    const ext = format.toLowerCase();
    if (ext === 'json') {
      return exportToJson(data, filePath);
    } else if (ext === 'csv') {
      return exportToCsv(data, filePath);
    } else if (ext === 'md' || ext === 'markdown') {
      return exportToMarkdown(data, filePath, title);
    }
    throw new Error(`Unsupported export format: ${format}`);
  }
}

module.exports = {
  RedditScraper,
};
